File: lc_soln/0347_top_k_frequent_elements/solution.py
from collections import Counter
from typing import List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = Counter(nums)
        num_and_freq = sorted(
            [(num, freq) for num, freq in counter.items()], key=lambda elem: -elem[1]
        )
        return [num_and_freq[i][0] for i in range(k)]


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0347_top_k_frequent_elements/solution.cs
public class Solution {
    public int[] TopKFrequent(int[] nums, int k) {
        var counter = new Dictionary<int, int>();
        foreach (int num in nums) {
            if (counter.ContainsKey(num)) {
                counter[num]++;
            } else {
                counter[num] = 1;
            }
        }
        var buckets = new Dictionary<int, List<int>>();
        foreach (KeyValuePair<int, int> kvp in counter) {
            if (!buckets.TryGetValue(kvp.Value, out List<int> list)) {
                list = new List<int>();
                buckets[kvp.Value] = list;
            }
            list.Add(kvp.Key);
        }
        var result = new List<int>();
        for (int i = nums.Count(); i > 0; --i) {
            if (buckets.TryGetValue(i, out List<int> bucket)) {
                foreach (int num in bucket) {
                    result.Add(num);
                    if (result.Count() == k) {
                        return result.ToArray();
                    }
                }
            }
        }
        return result.ToArray();
    }
}
